core/src/mock_alpenglow_consensus.rs

```diff
@@ -0,0 +1,741 @@
+                if should_exit.load(Ordering::Relaxed) {
+                    return;
+                }
+                continue;
+            };
+            let (slot, votor_msg) = match command {
+                Command::SendNotarize(slot) => (slot, VotorMessageType::Notarize),
+                Command::SendNotarizeCertificate(slot) => {
+                    (slot, VotorMessageType::NotarizeCertificate)
+                }
+                Command::SendFinalize(slot) => (slot, VotorMessageType::Finalize),
+                Command::SendFinalizeCertificate(slot) => {
+                    (slot, VotorMessageType::FinalizeCertificate)
+                }
+            };
+            let packet_size = (packet_size.load(Ordering::Relaxed) as usize)
+                .clamp(MOCK_VOTE_HEADER_SIZE, MOCK_VOTE_PACKET_MAX_SIZE);
+
+            prep_and_sign_packet(
+                &mut packet_buf,
+                slot,
+                &id,
+                votor_msg,
+                cluster_info.keypair().as_ref(),
+            );
+
+            // prepare addresses to send the packets
+            let mut send_instructions = Vec::with_capacity(2000);
+            {
+                let mut state = get_state_for_slot(state.as_slice(), slot).lock().unwrap();
+                // check if our task was aborted, avoid sending if it was.
+                if state.current_slot != slot {
+                    return;
+                }
+
+                for (peer, info) in state.peers.iter() {
+                    send_instructions.push((&packet_buf[0..packet_size], info.address));
+                    trace!(
+                        "{id}: send {votor_msg:?} for slot {slot} to {} for {peer}",
+                        info.address
+                    );
+                }
+            }
+            // broadcast to everybody at once
+            batch_send(&socket, send_instructions);
+        }
+    }
+
+    pub(crate) fn signal_new_slot(&mut self, slot: Slot, bank: &Bank) {
+        let Some(config) = get_test_config_from_account(bank) else {
+            debug!("Alpenglow votes disabled by on-chain config");
+            return;
+        };
+        // copy basic parameters from the config
+        self.packet_size
+            .store(config.packet_size, Ordering::Relaxed);
+        self.verify_signatures
+            .store(config.verify_signatures, Ordering::Relaxed);
+        let interval = config.test_interval_slots as u64;
+
+        // ensure we do not start process to early
+        if slot < self.highest_slot || slot < 3 {
+            trace!(
+                "Skipping AG logic for slot {slot}, current highest slot is {}",
+                self.highest_slot
+            );
+            return;
+        }
+
+        // If we fall too far behind and can not root banks, engage safety latch to stop the test
+        if interval > 0 && bank.slot() < slot - MAX_TOWER_HEIGHT {
```

Comment:
bank.slot() < slot - MAX_TOWER_HEIGHT seems a litlte aggressive since root bank is at least 32 slots behind the latest vote slot, possibly much more if there's any skipped slots. I feel like least a few validators will have this on every A2A blast.


Seems like we should use the same condition as vote-only mode MAX_ROOT_DISTANCE_FOR_VOTE_ONLY = 400

Reply:
The intention here was to prefer cluster health over test load. But if you are certain we can recover from 400 slots behind I'd obviously prefer to bump this up to ~50 slots.

Reply:
We are trying to turn our test off at the first sight of trouble. 400 is probably too brave? Does 100 work? I think Alex checked logs a few days ago and at the time it was ~40 on average.

Reply:
> in logs it was dead stuck at 32, occasionally going to 34-35

this was on actual testnet or on a test cluster?  There will be very little skips on a small internal cluster

Reply:
For reference, here's mean tower height for the last hour of testnet:
<img width="756" height="548" alt="image" src="https://github.com/user-attachments/assets/31840030-53a7-46ee-b39f-134fdeecb5f8" />

And looking at the last 24 hours on testnet, there are 7 unique validators that fell outside of this 132 slot tower height range at some point (root ~168-172 slots behind tip)

Reply:
100-200 seems like a good buffer :)